{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import yaml\n",
    "import chardet\n",
    "import ujson\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tax_placement(pident):\n",
    "    if pident >= tax_cutoffs['species']:\n",
    "        out = 'species'\n",
    "    elif pident >= tax_cutoffs['genus']:\n",
    "        out = 'genus'\n",
    "    elif pident >= tax_cutoffs['family']:\n",
    "        out = 'family'\n",
    "    elif pident >= tax_cutoffs['order']:\n",
    "        out = 'order'\n",
    "    elif pident < tax_cutoffs['order']:\n",
    "        out = 'class'\n",
    "    return out\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load Taxonomy File\n",
    "\n",
    "tax_file = '/vortexfs1/omics/alexander/data/mmetsp/reference_dir/taxonomy-table.txt'\n",
    "with open(tax_file, 'rb') as f:\n",
    "    result = chardet.detect(f.read())  # or readline if the file is large\n",
    "tax_table = pd.read_csv(tax_file, sep='\\t',encoding=result['encoding'])\n",
    "tax_table.columns = tax_table.columns.str.lower()\n",
    "tax_table=tax_table.set_index('source_id')\n",
    "#Read in tax-cutoffs\n",
    "with open(\"tax-cutoffs.yaml\", 'r') as stream:\n",
    "    tax_cutoffs = yaml.safe_load(stream)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cyverse_path</th>\n",
       "      <th>source_id_dup</th>\n",
       "      <th>ref_status</th>\n",
       "      <th>strain</th>\n",
       "      <th>supergroup</th>\n",
       "      <th>division</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>source_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>MMETSP0290</td>\n",
       "      <td>2225.0</td>\n",
       "      <td>MMETSP0290</td>\n",
       "      <td>Clean</td>\n",
       "      <td>CCMP2878</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Colpodellidea</td>\n",
       "      <td>Colpodellida</td>\n",
       "      <td>Chromeraceae</td>\n",
       "      <td>Chromera</td>\n",
       "      <td>Chromera velia</td>\n",
       "      <td>Curated by J. del Campo (PR2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MMETSP0288</td>\n",
       "      <td>2224.0</td>\n",
       "      <td>MMETSP0288</td>\n",
       "      <td>Clean</td>\n",
       "      <td>CCMP3155</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Colpodellidea</td>\n",
       "      <td>Vitrelladida</td>\n",
       "      <td>Vitrellaceae</td>\n",
       "      <td>Vitrella</td>\n",
       "      <td>Vitrella brassicaformis</td>\n",
       "      <td>Adl et al. 2019; Curated by J. del Campo and V...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MMETSP1451</td>\n",
       "      <td>2282.0</td>\n",
       "      <td>MMETSP1451</td>\n",
       "      <td>Clean</td>\n",
       "      <td>CCMP3346</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Colpodellidea</td>\n",
       "      <td>Vitrelladida</td>\n",
       "      <td>Vitrellaceae</td>\n",
       "      <td>Vitrella</td>\n",
       "      <td>Vitrella brassicaformis</td>\n",
       "      <td>Adl et al. 2019; Curated by J. del Campo and V...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MMETSP0372</td>\n",
       "      <td>1822.0</td>\n",
       "      <td>MMETSP0372</td>\n",
       "      <td>Clean</td>\n",
       "      <td>Grappler Inlet BC</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Gregarinomorphea</td>\n",
       "      <td>Eugregarinorida</td>\n",
       "      <td>Lecudinidae</td>\n",
       "      <td>Lankesteria</td>\n",
       "      <td>Lankesteria abbottii</td>\n",
       "      <td>Adl et al. 2019; Curated by J. del Campo (PR2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MMETSP0125</td>\n",
       "      <td>1740.0</td>\n",
       "      <td>MMETSP0125</td>\n",
       "      <td>Clean</td>\n",
       "      <td>ATCC 50986</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Ciliophora</td>\n",
       "      <td>Colpodea</td>\n",
       "      <td>Colpodea_X</td>\n",
       "      <td>Cyrtolophosidida</td>\n",
       "      <td>Aristerostoma</td>\n",
       "      <td>Aristerostoma sp.</td>\n",
       "      <td>Ciliophora EukRef curation Boscaro V.; previou...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MMETSP0990</td>\n",
       "      <td>1938.0</td>\n",
       "      <td>MMETSP0990</td>\n",
       "      <td>Clean</td>\n",
       "      <td>CCMP2098</td>\n",
       "      <td>Stramenopiles</td>\n",
       "      <td>Ochrophyta</td>\n",
       "      <td>Dictyochophyceae</td>\n",
       "      <td>Dictyochophyceae_X</td>\n",
       "      <td>Dictyochophyceae_XX</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>Not described; listed as Dictyochophyceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MMETSP0991</td>\n",
       "      <td>1943.0</td>\n",
       "      <td>MMETSP0991</td>\n",
       "      <td>Clean</td>\n",
       "      <td>CCMP2098</td>\n",
       "      <td>Stramenopiles</td>\n",
       "      <td>Ochrophyta</td>\n",
       "      <td>Dictyochophyceae</td>\n",
       "      <td>Dictyochophyceae_X</td>\n",
       "      <td>Dictyochophyceae_XX</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>Not described; listed as Dictyochophyceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MMETSP0992</td>\n",
       "      <td>1944.0</td>\n",
       "      <td>MMETSP0992</td>\n",
       "      <td>Clean</td>\n",
       "      <td>CCMP2098</td>\n",
       "      <td>Stramenopiles</td>\n",
       "      <td>Ochrophyta</td>\n",
       "      <td>Dictyochophyceae</td>\n",
       "      <td>Dictyochophyceae_X</td>\n",
       "      <td>Dictyochophyceae_XX</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>Not described; listed as Dictyochophyceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MMETSP0993</td>\n",
       "      <td>1945.0</td>\n",
       "      <td>MMETSP0993</td>\n",
       "      <td>Clean</td>\n",
       "      <td>CCMP2098</td>\n",
       "      <td>Stramenopiles</td>\n",
       "      <td>Ochrophyta</td>\n",
       "      <td>Dictyochophyceae</td>\n",
       "      <td>Dictyochophyceae_X</td>\n",
       "      <td>Dictyochophyceae_XX</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>Not described; listed as Dictyochophyceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MMETSP1433</td>\n",
       "      <td>2263.0</td>\n",
       "      <td>MMETSP1433</td>\n",
       "      <td>Clean</td>\n",
       "      <td>NY07348D</td>\n",
       "      <td>Stramenopiles</td>\n",
       "      <td>Stramenopiles_X</td>\n",
       "      <td>Stramenopiles_XX</td>\n",
       "      <td>Stramenopiles_XXX</td>\n",
       "      <td>Stramenopiles_XXXX</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>Not described</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>678 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            cyverse_path source_id_dup ref_status             strain  \\\n",
       "source_id                                                              \n",
       "MMETSP0290        2225.0    MMETSP0290      Clean           CCMP2878   \n",
       "MMETSP0288        2224.0    MMETSP0288      Clean           CCMP3155   \n",
       "MMETSP1451        2282.0    MMETSP1451      Clean           CCMP3346   \n",
       "MMETSP0372        1822.0    MMETSP0372      Clean  Grappler Inlet BC   \n",
       "MMETSP0125        1740.0    MMETSP0125      Clean         ATCC 50986   \n",
       "...                  ...           ...        ...                ...   \n",
       "MMETSP0990        1938.0    MMETSP0990      Clean           CCMP2098   \n",
       "MMETSP0991        1943.0    MMETSP0991      Clean           CCMP2098   \n",
       "MMETSP0992        1944.0    MMETSP0992      Clean           CCMP2098   \n",
       "MMETSP0993        1945.0    MMETSP0993      Clean           CCMP2098   \n",
       "MMETSP1433        2263.0    MMETSP1433      Clean           NY07348D   \n",
       "\n",
       "               supergroup         division             class  \\\n",
       "source_id                                                      \n",
       "MMETSP0290      Alveolata      Apicomplexa     Colpodellidea   \n",
       "MMETSP0288      Alveolata      Apicomplexa     Colpodellidea   \n",
       "MMETSP1451      Alveolata      Apicomplexa     Colpodellidea   \n",
       "MMETSP0372      Alveolata      Apicomplexa  Gregarinomorphea   \n",
       "MMETSP0125      Alveolata       Ciliophora          Colpodea   \n",
       "...                   ...              ...               ...   \n",
       "MMETSP0990  Stramenopiles       Ochrophyta  Dictyochophyceae   \n",
       "MMETSP0991  Stramenopiles       Ochrophyta  Dictyochophyceae   \n",
       "MMETSP0992  Stramenopiles       Ochrophyta  Dictyochophyceae   \n",
       "MMETSP0993  Stramenopiles       Ochrophyta  Dictyochophyceae   \n",
       "MMETSP1433  Stramenopiles  Stramenopiles_X  Stramenopiles_XX   \n",
       "\n",
       "                         order               family          genus  \\\n",
       "source_id                                                            \n",
       "MMETSP0290        Colpodellida         Chromeraceae       Chromera   \n",
       "MMETSP0288        Vitrelladida         Vitrellaceae       Vitrella   \n",
       "MMETSP1451        Vitrelladida         Vitrellaceae       Vitrella   \n",
       "MMETSP0372     Eugregarinorida          Lecudinidae    Lankesteria   \n",
       "MMETSP0125          Colpodea_X     Cyrtolophosidida  Aristerostoma   \n",
       "...                        ...                  ...            ...   \n",
       "MMETSP0990  Dictyochophyceae_X  Dictyochophyceae_XX      Uncertain   \n",
       "MMETSP0991  Dictyochophyceae_X  Dictyochophyceae_XX      Uncertain   \n",
       "MMETSP0992  Dictyochophyceae_X  Dictyochophyceae_XX      Uncertain   \n",
       "MMETSP0993  Dictyochophyceae_X  Dictyochophyceae_XX      Uncertain   \n",
       "MMETSP1433   Stramenopiles_XXX   Stramenopiles_XXXX      Uncertain   \n",
       "\n",
       "                            species  \\\n",
       "source_id                             \n",
       "MMETSP0290           Chromera velia   \n",
       "MMETSP0288  Vitrella brassicaformis   \n",
       "MMETSP1451  Vitrella brassicaformis   \n",
       "MMETSP0372     Lankesteria abbottii   \n",
       "MMETSP0125        Aristerostoma sp.   \n",
       "...                             ...   \n",
       "MMETSP0990                Uncertain   \n",
       "MMETSP0991                Uncertain   \n",
       "MMETSP0992                Uncertain   \n",
       "MMETSP0993                Uncertain   \n",
       "MMETSP1433                Uncertain   \n",
       "\n",
       "                                                        notes  \n",
       "source_id                                                      \n",
       "MMETSP0290                      Curated by J. del Campo (PR2)  \n",
       "MMETSP0288  Adl et al. 2019; Curated by J. del Campo and V...  \n",
       "MMETSP1451  Adl et al. 2019; Curated by J. del Campo and V...  \n",
       "MMETSP0372     Adl et al. 2019; Curated by J. del Campo (PR2)  \n",
       "MMETSP0125  Ciliophora EukRef curation Boscaro V.; previou...  \n",
       "...                                                       ...  \n",
       "MMETSP0990          Not described; listed as Dictyochophyceae  \n",
       "MMETSP0991          Not described; listed as Dictyochophyceae  \n",
       "MMETSP0992          Not described; listed as Dictyochophyceae  \n",
       "MMETSP0993          Not described; listed as Dictyochophyceae  \n",
       "MMETSP1433                                      Not described  \n",
       "\n",
       "[678 rows x 12 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tax_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/vortexfs1/omics/alexander/data/mmetsp/reference_dir/protein-table.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-6-30b6184983ed>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_cell_magic\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'timeit'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"prot_file = '/vortexfs1/omics/alexander/data/mmetsp/reference_dir/protein-table.txt'\\npdict = {}\\nwith open(prot_file, 'r')as f:\\n    for line in f:\\n        protein, mmetsp = line.split('\\\\t')\\n        pdict[protein]=mmetsp.strip()\\n\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/.conda/envs/EUKulele/lib/python3.7/site-packages/IPython/core/interactiveshell.py\u001b[0m in \u001b[0;36mrun_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2357\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuiltin_trap\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2358\u001b[0m                 \u001b[0margs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mmagic_arg_s\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2359\u001b[0;31m                 \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2360\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2361\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m</vortexfs1/home/halexander/.conda/envs/EUKulele/lib/python3.7/site-packages/decorator.py:decorator-gen-60>\u001b[0m in \u001b[0;36mtimeit\u001b[0;34m(self, line, cell, local_ns)\u001b[0m\n",
      "\u001b[0;32m~/.conda/envs/EUKulele/lib/python3.7/site-packages/IPython/core/magic.py\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(f, *a, **k)\u001b[0m\n\u001b[1;32m    185\u001b[0m     \u001b[0;31m# but it's overkill for just that one bit of state.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    186\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mmagic_deco\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 187\u001b[0;31m         \u001b[0mcall\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    188\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    189\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcallable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.conda/envs/EUKulele/lib/python3.7/site-packages/IPython/core/magics/execution.py\u001b[0m in \u001b[0;36mtimeit\u001b[0;34m(self, line, cell, local_ns)\u001b[0m\n\u001b[1;32m   1156\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mindex\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1157\u001b[0m                 \u001b[0mnumber\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m10\u001b[0m \u001b[0;34m**\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1158\u001b[0;31m                 \u001b[0mtime_number\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtimer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtimeit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnumber\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1159\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mtime_number\u001b[0m \u001b[0;34m>=\u001b[0m \u001b[0;36m0.2\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1160\u001b[0m                     \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.conda/envs/EUKulele/lib/python3.7/site-packages/IPython/core/magics/execution.py\u001b[0m in \u001b[0;36mtimeit\u001b[0;34m(self, number)\u001b[0m\n\u001b[1;32m    167\u001b[0m         \u001b[0mgc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdisable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    168\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 169\u001b[0;31m             \u001b[0mtiming\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minner\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mit\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtimer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    170\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    171\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mgcold\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<magic-timeit>\u001b[0m in \u001b[0;36minner\u001b[0;34m(_it, _timer)\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/vortexfs1/omics/alexander/data/mmetsp/reference_dir/protein-table.txt'"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "prot_file = '/vortexfs1/omics/alexander/data/mmetsp/reference_dir/protein-table.txt'\n",
    "pdict = {}\n",
    "with open(prot_file, 'r')as f:\n",
    "    for line in f:\n",
    "        protein, mmetsp = line.split('\\t')\n",
    "        pdict[protein]=mmetsp.strip()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_table = 'output/MAGs/diamond/IO-all-DCM-0-8-5-00_bin-230.diamond.out'\n",
    "df = pd.read_csv(output_table, sep = '\\t', header = None)\n",
    "df.columns = ['qseqid', 'sseqid', 'pident', 'length', 'mismatch', 'gapopen', 'qstart', 'qend', 'sstart', 'send', 'evalue', 'bitscore']\n",
    "df['ssqidMMETSP']=df.sseqid.map(imp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For MetaTranscriptomes\n",
    "classification_df = pd.DataFrame(columns = ['classification_level', 'classification', 'max_pid'])\n",
    "for t,dd in df.groupby('qseqid'):\n",
    "    md = dd.pident.max()\n",
    "    ds = list(set(dd[dd.pident==md]['ssqidMMETSP']))\n",
    "    assignment = tax_placement(md)\n",
    "    if len(ds)==1:\n",
    "        classification = tax_table.loc[ds[0], assignment]\n",
    "    else:\n",
    "        classification_0 = []\n",
    "        for d in ds:\n",
    "            classification_0.append(tax_table.loc[ds[0], assignment])\n",
    "            if len(set(classification_0)) ==1:\n",
    "                classification = list(set(classification_0))[0]\n",
    "            else:\n",
    "                pass\n",
    "    classification_df.loc[t] = [assignment, classification, md]\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>classification_level</th>\n",
       "      <th>full_classification</th>\n",
       "      <th>classification</th>\n",
       "      <th>max_pid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>g1[IO-all-DCM-0-8-5-00_k119_10027296:607-4321]</td>\n",
       "      <td>class</td>\n",
       "      <td>Stramenopiles; Opalozoa; Bicoecea</td>\n",
       "      <td>Bicoecea</td>\n",
       "      <td>39.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>g1[IO-all-DCM-0-8-5-00_k119_10028232:1-797]</td>\n",
       "      <td>class</td>\n",
       "      <td>Hacrobia; Cryptophyta; Cryptophyceae</td>\n",
       "      <td>Cryptophyceae</td>\n",
       "      <td>25.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>g1[IO-all-DCM-0-8-5-00_k119_10033933:4588-13506]</td>\n",
       "      <td>order</td>\n",
       "      <td>Stramenopiles; Ochrophyta; Raphidophyceae; Rap...</td>\n",
       "      <td>Raphidophyceae_X</td>\n",
       "      <td>52.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>g1[IO-all-DCM-0-8-5-00_k119_10044618:374-4385]</td>\n",
       "      <td>class</td>\n",
       "      <td>Stramenopiles; Ochrophyta; Pelagophyceae</td>\n",
       "      <td>Pelagophyceae</td>\n",
       "      <td>43.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>g1[IO-all-DCM-0-8-5-00_k119_10051159:95-6545]</td>\n",
       "      <td>class</td>\n",
       "      <td>Archaeplastida; Chlorophyta; Chlorodendrophyceae</td>\n",
       "      <td>Chlorodendrophyceae</td>\n",
       "      <td>37.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>g4[IO-all-DCM-0-8-5-00_k119_16450953:4064-4594]</td>\n",
       "      <td>order</td>\n",
       "      <td>Alveolata; Dinoflagellata; Dinophyceae; Gonyau...</td>\n",
       "      <td>Gonyaulacales</td>\n",
       "      <td>61.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>g4[IO-all-DCM-0-8-5-00_k119_2469638:7354-7692]</td>\n",
       "      <td>order</td>\n",
       "      <td>Stramenopiles; Opalozoa; Bicoecea; Anoecales</td>\n",
       "      <td>Anoecales</td>\n",
       "      <td>54.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>g4[IO-all-DCM-0-8-5-00_k119_4804390:4241-5176]</td>\n",
       "      <td>class</td>\n",
       "      <td>Stramenopiles; Ochrophyta; Bacillariophyta</td>\n",
       "      <td>Bacillariophyta</td>\n",
       "      <td>36.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>g4[IO-all-DCM-0-8-5-00_k119_7460344:12265-14439]</td>\n",
       "      <td>class</td>\n",
       "      <td>Stramenopiles; Ochrophyta; Raphidophyceae</td>\n",
       "      <td>Raphidophyceae</td>\n",
       "      <td>47.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>g4[IO-all-DCM-0-8-5-00_k119_7690977:6134-7549]</td>\n",
       "      <td>order</td>\n",
       "      <td>Alveolata; Dinoflagellata; Dinophyceae; Gonyau...</td>\n",
       "      <td>Gonyaulacales</td>\n",
       "      <td>55.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>433 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 classification_level  \\\n",
       "g1[IO-all-DCM-0-8-5-00_k119_10027296:607-4321]                  class   \n",
       "g1[IO-all-DCM-0-8-5-00_k119_10028232:1-797]                     class   \n",
       "g1[IO-all-DCM-0-8-5-00_k119_10033933:4588-13506]                order   \n",
       "g1[IO-all-DCM-0-8-5-00_k119_10044618:374-4385]                  class   \n",
       "g1[IO-all-DCM-0-8-5-00_k119_10051159:95-6545]                   class   \n",
       "...                                                               ...   \n",
       "g4[IO-all-DCM-0-8-5-00_k119_16450953:4064-4594]                 order   \n",
       "g4[IO-all-DCM-0-8-5-00_k119_2469638:7354-7692]                  order   \n",
       "g4[IO-all-DCM-0-8-5-00_k119_4804390:4241-5176]                  class   \n",
       "g4[IO-all-DCM-0-8-5-00_k119_7460344:12265-14439]                class   \n",
       "g4[IO-all-DCM-0-8-5-00_k119_7690977:6134-7549]                  order   \n",
       "\n",
       "                                                                                full_classification  \\\n",
       "g1[IO-all-DCM-0-8-5-00_k119_10027296:607-4321]                    Stramenopiles; Opalozoa; Bicoecea   \n",
       "g1[IO-all-DCM-0-8-5-00_k119_10028232:1-797]                    Hacrobia; Cryptophyta; Cryptophyceae   \n",
       "g1[IO-all-DCM-0-8-5-00_k119_10033933:4588-13506]  Stramenopiles; Ochrophyta; Raphidophyceae; Rap...   \n",
       "g1[IO-all-DCM-0-8-5-00_k119_10044618:374-4385]             Stramenopiles; Ochrophyta; Pelagophyceae   \n",
       "g1[IO-all-DCM-0-8-5-00_k119_10051159:95-6545]      Archaeplastida; Chlorophyta; Chlorodendrophyceae   \n",
       "...                                                                                             ...   \n",
       "g4[IO-all-DCM-0-8-5-00_k119_16450953:4064-4594]   Alveolata; Dinoflagellata; Dinophyceae; Gonyau...   \n",
       "g4[IO-all-DCM-0-8-5-00_k119_2469638:7354-7692]         Stramenopiles; Opalozoa; Bicoecea; Anoecales   \n",
       "g4[IO-all-DCM-0-8-5-00_k119_4804390:4241-5176]           Stramenopiles; Ochrophyta; Bacillariophyta   \n",
       "g4[IO-all-DCM-0-8-5-00_k119_7460344:12265-14439]          Stramenopiles; Ochrophyta; Raphidophyceae   \n",
       "g4[IO-all-DCM-0-8-5-00_k119_7690977:6134-7549]    Alveolata; Dinoflagellata; Dinophyceae; Gonyau...   \n",
       "\n",
       "                                                       classification  max_pid  \n",
       "g1[IO-all-DCM-0-8-5-00_k119_10027296:607-4321]               Bicoecea     39.3  \n",
       "g1[IO-all-DCM-0-8-5-00_k119_10028232:1-797]             Cryptophyceae     25.9  \n",
       "g1[IO-all-DCM-0-8-5-00_k119_10033933:4588-13506]     Raphidophyceae_X     52.9  \n",
       "g1[IO-all-DCM-0-8-5-00_k119_10044618:374-4385]          Pelagophyceae     43.8  \n",
       "g1[IO-all-DCM-0-8-5-00_k119_10051159:95-6545]     Chlorodendrophyceae     37.4  \n",
       "...                                                               ...      ...  \n",
       "g4[IO-all-DCM-0-8-5-00_k119_16450953:4064-4594]         Gonyaulacales     61.5  \n",
       "g4[IO-all-DCM-0-8-5-00_k119_2469638:7354-7692]              Anoecales     54.5  \n",
       "g4[IO-all-DCM-0-8-5-00_k119_4804390:4241-5176]        Bacillariophyta     36.8  \n",
       "g4[IO-all-DCM-0-8-5-00_k119_7460344:12265-14439]       Raphidophyceae     47.9  \n",
       "g4[IO-all-DCM-0-8-5-00_k119_7690977:6134-7549]          Gonyaulacales     55.2  \n",
       "\n",
       "[433 rows x 4 columns]"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# For MetaGenomes\n",
    "level_dict = {'class':['supergroup','division','class'],\n",
    "                       'order':['supergroup','division','class', 'order'],\n",
    "                       'family':['supergroup','division','class', 'order', 'family'],\n",
    "                       'genus': ['supergroup','division','class', 'order', 'family','genus'],\n",
    "                        'species':['supergroup','division','class', 'order', 'family','genus', 'species']}\n",
    "classification_df = pd.DataFrame(columns = ['classification_level', 'full_classification','classification', 'max_pid'])\n",
    "for t,dd in df.groupby('qseqid'):\n",
    "    md = dd.pident.max()\n",
    "    ds = list(set(dd[dd.pident==md]['ssqidMMETSP']))\n",
    "    assignment = tax_placement(md)\n",
    "    full_assignment = level_dict[assignment]\n",
    "    if len(ds)==1:\n",
    "        classification = tax_table.loc[ds[0], assignment]\n",
    "        full_classification = tax_table.loc[ds[0], full_assignment]\n",
    "        full_classification='; '.join(list(full_classification))\n",
    "    else:\n",
    "        classification_0 = []\n",
    "        full_classification_0=[]\n",
    "        for d in ds:\n",
    "            classification_0.append(tax_table.loc[ds[0], assignment])\n",
    "            fc = tax_table.loc[ds[0], full_assignment]\n",
    "            full_classification_0.append('; '.join(list(fc)))\n",
    "            if len(set(classification_0)) ==1:\n",
    "                classification = list(set(classification_0))[0]\n",
    "                full_classification=list(set(full_classification_0))[0]\n",
    "            else:\n",
    "                pass\n",
    "    classification_df.loc[t] = [assignment, full_classification, classification, md]\n",
    "classification_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "levels = ['supergroup','division','class','order','family','genus','species']\n",
    "tmp_df = pd.DataFrame(classification_df.full_classification.str.split('; '))\n",
    "for cl in tmp_df.index:\n",
    "    lineage = tmp_df.loc[cl,'full_classification']\n",
    "    for i,l in enumerate(levels):\n",
    "        if len(lineage)>i:\n",
    "            outdf.loc[cl,l]=lineage[i]\n",
    "        else:\n",
    "            outdf.loc[cl,l]=np.nan\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "tax_dict = {}\n",
    "total_len=len(outdf)\n",
    "for l in levels:\n",
    "    tax_dict[l]={}\n",
    "    column_sum = outdf.groupby(l)[l].sum()\n",
    "    column_count = outdf.groupby(l)[l].count()\n",
    "    norm_count=column_count/total_len\n",
    "    tax_dict[l]=norm_count\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_df = pd.DataFrame(index=levels, columns=['max_taxa','percent_id'])\n",
    "for key in tax_dict:\n",
    "    highest_tax = tax_dict[key][tax_dict[key]==tax_dict[key].max()].index\n",
    "    max_val = tax_dict[key].max()\n",
    "    max_df.loc[key]=highest_tax[0], max_val\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [],
   "source": [
    "outdf = pd.read_csv('output/MAGs/SAO-all-SRF-0-8-5-00_bin-59-estimated-taxonomy.out', sep='\\t', index_col=0)\n",
    "\n",
    "levels = ['supergroup','division','class','order','family','genus','species']\n",
    "tmp_df = pd.DataFrame(outdf.full_classification.str.split('; '))\n",
    "for cl in tmp_df.index:\n",
    "    lineage = tmp_df.loc[cl,'full_classification']\n",
    "    for i,l in enumerate(levels):\n",
    "        if len(lineage)>i:\n",
    "            outdf.loc[cl,l]=lineage[i]\n",
    "        else:\n",
    "            outdf.loc[cl,l]=np.nan\n",
    "\n",
    "tax_dict = {}\n",
    "total_len=len(outdf)\n",
    "for l in levels:\n",
    "    tax_dict[l]={}\n",
    "    column_sum = outdf.groupby(l)[l].sum()\n",
    "    column_count = outdf.groupby(l)[l].count()\n",
    "    norm_count=column_count/total_len\n",
    "    tax_dict[l]=norm_count\n",
    "\n",
    "    \n",
    "max_df = pd.DataFrame(index=levels, columns=['max_taxa','percent_id'])\n",
    "for key in tax_dict:\n",
    "    highest_tax = tax_dict[key][tax_dict[key]==tax_dict[key].max()].index\n",
    "    max_val = tax_dict[key].max()\n",
    "    max_df.loc[key]=highest_tax[0], max_val\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>max_taxa</th>\n",
       "      <th>percent_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>supergroup</td>\n",
       "      <td>Stramenopiles</td>\n",
       "      <td>0.547887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>division</td>\n",
       "      <td>Ochrophyta</td>\n",
       "      <td>0.457746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>class</td>\n",
       "      <td>Bacillariophyta</td>\n",
       "      <td>0.11831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>order</td>\n",
       "      <td>Raphidophyceae_X</td>\n",
       "      <td>0.0633803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>family</td>\n",
       "      <td>Raphidophyceae_XX</td>\n",
       "      <td>0.0183099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>genus</td>\n",
       "      <td>Florenciella</td>\n",
       "      <td>0.0028169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>species</td>\n",
       "      <td>Coccolithus pelagicus ssp braarudi</td>\n",
       "      <td>0.00140845</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      max_taxa  percent_id\n",
       "supergroup                       Stramenopiles    0.547887\n",
       "division                            Ochrophyta    0.457746\n",
       "class                          Bacillariophyta     0.11831\n",
       "order                         Raphidophyceae_X   0.0633803\n",
       "family                       Raphidophyceae_XX   0.0183099\n",
       "genus                             Florenciella   0.0028169\n",
       "species     Coccolithus pelagicus ssp braarudi  0.00140845"
      ]
     },
     "execution_count": 175,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "for t,dd in df.groupby('qseqid'):\n",
    "    dd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "mfi = '/vortexfs1/omics/alexander/data/mmetsp/reference_dir/protein-species-map.json'\n",
    "with open(mfi, 'rb') as fp:\n",
    "    imp = ujson.load(fp)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cyverse_path</th>\n",
       "      <th>source_id_dup</th>\n",
       "      <th>ref_status</th>\n",
       "      <th>strain</th>\n",
       "      <th>supergroup</th>\n",
       "      <th>division</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>source_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>MMETSP0290</td>\n",
       "      <td>2225.0</td>\n",
       "      <td>MMETSP0290</td>\n",
       "      <td>Clean</td>\n",
       "      <td>CCMP2878</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Colpodellidea</td>\n",
       "      <td>Colpodellida</td>\n",
       "      <td>Chromeraceae</td>\n",
       "      <td>Chromera</td>\n",
       "      <td>Chromera velia</td>\n",
       "      <td>Curated by J. del Campo (PR2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MMETSP0288</td>\n",
       "      <td>2224.0</td>\n",
       "      <td>MMETSP0288</td>\n",
       "      <td>Clean</td>\n",
       "      <td>CCMP3155</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Colpodellidea</td>\n",
       "      <td>Vitrelladida</td>\n",
       "      <td>Vitrellaceae</td>\n",
       "      <td>Vitrella</td>\n",
       "      <td>Vitrella brassicaformis</td>\n",
       "      <td>Adl et al. 2019; Curated by J. del Campo and V...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MMETSP1451</td>\n",
       "      <td>2282.0</td>\n",
       "      <td>MMETSP1451</td>\n",
       "      <td>Clean</td>\n",
       "      <td>CCMP3346</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Colpodellidea</td>\n",
       "      <td>Vitrelladida</td>\n",
       "      <td>Vitrellaceae</td>\n",
       "      <td>Vitrella</td>\n",
       "      <td>Vitrella brassicaformis</td>\n",
       "      <td>Adl et al. 2019; Curated by J. del Campo and V...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MMETSP0372</td>\n",
       "      <td>1822.0</td>\n",
       "      <td>MMETSP0372</td>\n",
       "      <td>Clean</td>\n",
       "      <td>Grappler Inlet BC</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>Gregarinomorphea</td>\n",
       "      <td>Eugregarinorida</td>\n",
       "      <td>Lecudinidae</td>\n",
       "      <td>Lankesteria</td>\n",
       "      <td>Lankesteria abbottii</td>\n",
       "      <td>Adl et al. 2019; Curated by J. del Campo (PR2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MMETSP0125</td>\n",
       "      <td>1740.0</td>\n",
       "      <td>MMETSP0125</td>\n",
       "      <td>Clean</td>\n",
       "      <td>ATCC 50986</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Ciliophora</td>\n",
       "      <td>Colpodea</td>\n",
       "      <td>Colpodea_X</td>\n",
       "      <td>Cyrtolophosidida</td>\n",
       "      <td>Aristerostoma</td>\n",
       "      <td>Aristerostoma sp.</td>\n",
       "      <td>Ciliophora EukRef curation Boscaro V.; previou...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MMETSP0990</td>\n",
       "      <td>1938.0</td>\n",
       "      <td>MMETSP0990</td>\n",
       "      <td>Clean</td>\n",
       "      <td>CCMP2098</td>\n",
       "      <td>Stramenopiles</td>\n",
       "      <td>Ochrophyta</td>\n",
       "      <td>Dictyochophyceae</td>\n",
       "      <td>Dictyochophyceae_X</td>\n",
       "      <td>Dictyochophyceae_XX</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>Not described; listed as Dictyochophyceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MMETSP0991</td>\n",
       "      <td>1943.0</td>\n",
       "      <td>MMETSP0991</td>\n",
       "      <td>Clean</td>\n",
       "      <td>CCMP2098</td>\n",
       "      <td>Stramenopiles</td>\n",
       "      <td>Ochrophyta</td>\n",
       "      <td>Dictyochophyceae</td>\n",
       "      <td>Dictyochophyceae_X</td>\n",
       "      <td>Dictyochophyceae_XX</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>Not described; listed as Dictyochophyceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MMETSP0992</td>\n",
       "      <td>1944.0</td>\n",
       "      <td>MMETSP0992</td>\n",
       "      <td>Clean</td>\n",
       "      <td>CCMP2098</td>\n",
       "      <td>Stramenopiles</td>\n",
       "      <td>Ochrophyta</td>\n",
       "      <td>Dictyochophyceae</td>\n",
       "      <td>Dictyochophyceae_X</td>\n",
       "      <td>Dictyochophyceae_XX</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>Not described; listed as Dictyochophyceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MMETSP0993</td>\n",
       "      <td>1945.0</td>\n",
       "      <td>MMETSP0993</td>\n",
       "      <td>Clean</td>\n",
       "      <td>CCMP2098</td>\n",
       "      <td>Stramenopiles</td>\n",
       "      <td>Ochrophyta</td>\n",
       "      <td>Dictyochophyceae</td>\n",
       "      <td>Dictyochophyceae_X</td>\n",
       "      <td>Dictyochophyceae_XX</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>Not described; listed as Dictyochophyceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MMETSP1433</td>\n",
       "      <td>2263.0</td>\n",
       "      <td>MMETSP1433</td>\n",
       "      <td>Clean</td>\n",
       "      <td>NY07348D</td>\n",
       "      <td>Stramenopiles</td>\n",
       "      <td>Stramenopiles_X</td>\n",
       "      <td>Stramenopiles_XX</td>\n",
       "      <td>Stramenopiles_XXX</td>\n",
       "      <td>Stramenopiles_XXXX</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>Not described</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>678 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            cyverse_path source_id_dup ref_status             strain  \\\n",
       "source_id                                                              \n",
       "MMETSP0290        2225.0    MMETSP0290      Clean           CCMP2878   \n",
       "MMETSP0288        2224.0    MMETSP0288      Clean           CCMP3155   \n",
       "MMETSP1451        2282.0    MMETSP1451      Clean           CCMP3346   \n",
       "MMETSP0372        1822.0    MMETSP0372      Clean  Grappler Inlet BC   \n",
       "MMETSP0125        1740.0    MMETSP0125      Clean         ATCC 50986   \n",
       "...                  ...           ...        ...                ...   \n",
       "MMETSP0990        1938.0    MMETSP0990      Clean           CCMP2098   \n",
       "MMETSP0991        1943.0    MMETSP0991      Clean           CCMP2098   \n",
       "MMETSP0992        1944.0    MMETSP0992      Clean           CCMP2098   \n",
       "MMETSP0993        1945.0    MMETSP0993      Clean           CCMP2098   \n",
       "MMETSP1433        2263.0    MMETSP1433      Clean           NY07348D   \n",
       "\n",
       "               supergroup         division             class  \\\n",
       "source_id                                                      \n",
       "MMETSP0290      Alveolata      Apicomplexa     Colpodellidea   \n",
       "MMETSP0288      Alveolata      Apicomplexa     Colpodellidea   \n",
       "MMETSP1451      Alveolata      Apicomplexa     Colpodellidea   \n",
       "MMETSP0372      Alveolata      Apicomplexa  Gregarinomorphea   \n",
       "MMETSP0125      Alveolata       Ciliophora          Colpodea   \n",
       "...                   ...              ...               ...   \n",
       "MMETSP0990  Stramenopiles       Ochrophyta  Dictyochophyceae   \n",
       "MMETSP0991  Stramenopiles       Ochrophyta  Dictyochophyceae   \n",
       "MMETSP0992  Stramenopiles       Ochrophyta  Dictyochophyceae   \n",
       "MMETSP0993  Stramenopiles       Ochrophyta  Dictyochophyceae   \n",
       "MMETSP1433  Stramenopiles  Stramenopiles_X  Stramenopiles_XX   \n",
       "\n",
       "                         order               family          genus  \\\n",
       "source_id                                                            \n",
       "MMETSP0290        Colpodellida         Chromeraceae       Chromera   \n",
       "MMETSP0288        Vitrelladida         Vitrellaceae       Vitrella   \n",
       "MMETSP1451        Vitrelladida         Vitrellaceae       Vitrella   \n",
       "MMETSP0372     Eugregarinorida          Lecudinidae    Lankesteria   \n",
       "MMETSP0125          Colpodea_X     Cyrtolophosidida  Aristerostoma   \n",
       "...                        ...                  ...            ...   \n",
       "MMETSP0990  Dictyochophyceae_X  Dictyochophyceae_XX      Uncertain   \n",
       "MMETSP0991  Dictyochophyceae_X  Dictyochophyceae_XX      Uncertain   \n",
       "MMETSP0992  Dictyochophyceae_X  Dictyochophyceae_XX      Uncertain   \n",
       "MMETSP0993  Dictyochophyceae_X  Dictyochophyceae_XX      Uncertain   \n",
       "MMETSP1433   Stramenopiles_XXX   Stramenopiles_XXXX      Uncertain   \n",
       "\n",
       "                            species  \\\n",
       "source_id                             \n",
       "MMETSP0290           Chromera velia   \n",
       "MMETSP0288  Vitrella brassicaformis   \n",
       "MMETSP1451  Vitrella brassicaformis   \n",
       "MMETSP0372     Lankesteria abbottii   \n",
       "MMETSP0125        Aristerostoma sp.   \n",
       "...                             ...   \n",
       "MMETSP0990                Uncertain   \n",
       "MMETSP0991                Uncertain   \n",
       "MMETSP0992                Uncertain   \n",
       "MMETSP0993                Uncertain   \n",
       "MMETSP1433                Uncertain   \n",
       "\n",
       "                                                        notes  \n",
       "source_id                                                      \n",
       "MMETSP0290                      Curated by J. del Campo (PR2)  \n",
       "MMETSP0288  Adl et al. 2019; Curated by J. del Campo and V...  \n",
       "MMETSP1451  Adl et al. 2019; Curated by J. del Campo and V...  \n",
       "MMETSP0372     Adl et al. 2019; Curated by J. del Campo (PR2)  \n",
       "MMETSP0125  Ciliophora EukRef curation Boscaro V.; previou...  \n",
       "...                                                       ...  \n",
       "MMETSP0990          Not described; listed as Dictyochophyceae  \n",
       "MMETSP0991          Not described; listed as Dictyochophyceae  \n",
       "MMETSP0992          Not described; listed as Dictyochophyceae  \n",
       "MMETSP0993          Not described; listed as Dictyochophyceae  \n",
       "MMETSP1433                                      Not described  \n",
       "\n",
       "[678 rows x 12 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tax_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-EUKulele]",
   "language": "python",
   "name": "conda-env-.conda-EUKulele-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
